runtime-modules/project-token/src/lib.rs

```diff
@@ -800,6 +801,239 @@ decl_module! {
+                            user_amount,
+                            // No restrictions on repayable bloat bond,
+                            // since only usable balance is allowed
+                            RepayableBloatBond::new(bloat_bond, None)
+                    );
+                Self::do_insert_new_account_for_token(token_id, &member_id, new_account_info);
+                Self::transfer_joy(&sender, &Self::module_treasury_account(), bloat_bond)?;
+            } else {
+                AccountInfoByTokenAndMember::<T>::mutate(token_id, member_id, |account_data| {
+                    account_data.amount = account_data.amount.saturating_add(user_amount);
+                });
+            }
+
+            TokenInfoById::<T>::mutate(token_id, |token_data| {
+                token_data.total_supply = token_data.total_supply.saturating_add(amount);
+                token_data.tokens_issued = token_data.tokens_issued.saturating_add(amount);
+            });
+            Self::transfer_joy(&sender, &amm_reserve_account, price)?;
+
+            Self::deposit_event(RawEvent::TokenBonded(token_id, member_id, amount, price));
+
+            Ok(())
+        }
+
+        /// Mint token and transfer joys to the AMM
+        /// Preconditions
+        /// - origin, member_id pair must be a valid authentication pair
+        /// - token_id must exist
+        /// - token_id, member_id must be valid account coordinates
+        /// - user usable CRT balance must be at least `amount`
+        /// - deadline constraint respected if provided
+        /// - slippage tolerance constraints respected if provided
+        /// - token total supply and amount value must be s.t. `eval` function doesn't overflow
+        /// - amm treasury acconut must have sufficient JOYs for the operation
+        ///
+        /// Postconditions
+        /// - `amount` burn from user account
+        /// - total supply decreased by amount
+        /// - respective JOY amount transferred from amm treasury account to user account
+        /// - event deposited
+        #[weight = 100_000_000] // TODO: adjust weight
+        fn unbond(origin, token_id: T::TokenId, member_id: T::MemberId, amount: <T as Config>::Balance, deadline: Option<<T as timestamp::Config>::Moment>, slippage_tolerance: Option<(Permill, JoyBalanceOf<T>)>) -> DispatchResult {
+            if amount.is_zero() {
+                return Ok(()); // noop
+            }
+
+            let sender = ensure_signed(origin.clone())?;
+
+            T::MemberOriginValidator::ensure_member_controller_account_origin(
+                origin,
+                member_id
+            )?;
+
+            let token_data = Self::ensure_token_exists(token_id)?;
+            let curve = token_data.bonding_curve.ok_or(Error::<T>::NotInAmmState)?;
+            let user_acc_data = Self::ensure_account_data_exists(token_id, &member_id)?;
+
+            ensure!(user_acc_data.amount >= amount, Error::<T>::UnsufficientTokenAmount);
```

Comment:
To prevent vesting funds leakage you should check `transferrable` balance here:
```suggestion
            ensure!(
                user_acc_data.transferrable::<T>(Self::current_block()) >= amount,
                Error::<T>::InsufficientTokenAmount
            );
```